RFCs/0031-Multi-Test-Run-Finalization.md

```diff
@@ -0,0 +1,137 @@
+        /// <param name="logger">Message logger</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        /// <returns>Attachments after reprocessing</returns>
+        ICollection<AttachmentSet> HandleDataCollectionAttachmentSets(ICollection<AttachmentSet> attachments, IProgress<int> progressReporter, IMessageLogger logger, CancellationToken cancellationToken);
+    }
+}
+```
+
+Method `GetExtensionUris` should provide all Uris for data attachments which are handled by current collector. Test platform will provide to collector only attachments with such Uris. It's also required that result of method `HandleDataCollectionAttachmentSets` will contain only attachments with such Uris.
+
+`IsAssociative` should indicate if `HandleDataCollectionAttachmentSets` is [associative](https://en.wikipedia.org/wiki/Associative_property).
+
+If `IsAssociative` is `True` Test Platform may try to speed up executions/merging by combining/merging data attachments as soon as possible when any two test executions are done. For example let's assume we have 5 test executions which are generating 5 data attachments: `a1`, `a2`, `a3`, `a4` and `a5`. Test platform could perform invocations:
+* `var result1 = HandleDataCollectionAttachmentSets([a1, a2, a3], ...);` when first 3 executions are done
+* `var result2 = HandleDataCollectionAttachmentSets(result1.Concat([a4]), ...);` when 4th execution is done
+* `var finalResult = HandleDataCollectionAttachmentSets(result2.Concat([a5]), ...);` when last test execution is done
+
+If `IsAssociative` is `False` then Test Platform will wait for all test executions to finish and call `HandleDataCollectionAttachmentSets` only once:
+* `var finalResult = HandleDataCollectionAttachmentSets([a1, a2, a3, a4, a5], ...);`
+
+By default `IsAssociative` should be `False`, unless processing can take longer time and it's beneficial to start handling as soon as possible.
+
+
+
+2. Introduce a new `FinalizeMultiTestRunAsync` method in [IVsTestConsoleWrapper](https://github.com/microsoft/vstest/blob/master/src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs) interface:
+
+```
+/// <summary>
+/// Provides back all attachments to TestPlatform for additional processing (for example merging)
+/// </summary>
+/// <param name="attachments">Collection of attachments</param>
+/// <param name="multiTestRunCompleted">Indicates that all test executions are done and all data is provided</param>
+/// <param name="collectMetrics">Enables metrics collection</param>
+/// <param name="multiTestRunFinalizationCompleteEventsHandler">EventHandler to receive session complete event</param>
+/// <param name="cancellationToken">Cancellation token</param>        
+Task FinalizeMultiTestRunAsync(IEnumerable<AttachmentSet> attachments, bool multiTestRunCompleted, bool collectMetrics, IMultiTestRunFinalizationEventsHandler eventsHandler, CancellationToken cancellationToken);
```

Comment:
We should also mention the raw json messages passed through to vstest.console so any clients talking through json (and not using the managed translation layer) can benefit from this as well.

Reply:
Documented those messages, but do we really want to have such data inside RFC? How we manage consistency? Can we replace it with links to code when we merge this to master?

Reply:
Pointing to source code with a fixed SHA so it doesn't change sounds good. For the protocol, I was referring to a something similar to this: [https://github.com/microsoft/vstest-docs/blob/master/RFCs/0012-Editors-API-RevisionUpdate-15.4.md](https://github.com/microsoft/vstest-docs/blob/master/RFCs/0012-Editors-API-RevisionUpdate-15.4.md). We could put that out into a separate doc to make it more discoverable.

---
In reply to: [446987820](https://github.com/microsoft/vstest-docs/pull/235#discussion_r446987820) [](ancestors = 446987820)

Reply:
Created separate task. Its not changing interfaces so we can do it later.